### DISCUSSION TITLE
Rule Engine Odd Behavior

### DISCUSSION THREAD

**Discussion by sherif-fanous** [firefly-iii/firefly-iii#6096] [Firefly III]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely me, not you](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

I have created multiple rules using the `Create rule based on transaction` action. When I try to test the rules using the `See matching transactions` option the query returns 0 matches!!!

I've spent the last hour trying to debug this issue and went through multiple iterations of simplifying the rule until I got a match but the behaivor just seems buggy in my opinion.

These are the initial triggers I had for the rule

`transaction_type:"Withdrawal" currency_is:"Canadian dollar (CAD)" description_is:"MCAP SERVICE     MTG" source_account_is:"TD Canada Trust  All-Inclusive Chequing" destination_account_is:"(no name)"`

This is the rule I managed to get working

`transaction_type:"Withdrawal" currency_is:"Canadian dollar (CAD)" description_starts:"MCAP SERVICE" destination_account_is:"(no name)"`

So for some reason `source_account_is` fails to match although all transactions are in this account, so even such a query will return 0 matches

`transaction_type:"Withdrawal" currency_is:"Canadian dollar (CAD)" source_account_is:"TD Canada Trust  All-Inclusive Chequing" destination_account_is:"(no name)"`

Also more strangely is how the description matching works.
If I try `description_starts:"MCAP SERVICE "` (Note the trailing space) the query still returns the expected matches.

I must have done about 20 tests where on each iteration I add an additional trailing space and the query still returns matches!!!

`transaction_type:"Withdrawal" currency_is:"Canadian dollar (CAD)" description_starts:"MCAP SERVICE                     " destination_account_is:"(no name)"`

Looking at the logs, it seems description_starts trims successive spaces to just 1 space regardless.

There is no way I could get `description_is` to work though. I thought there might be special characters in the string that were preventing the match but I checked with Python and all the characters are either English letters or a space (ord = 32)

### Expected behaviour

Query should return matching transactions.

### Debug information

```Debug information generated at 2022-05-23 00:00:26 America/Toronto for Firefly III version **5.7.6**.

| Scope | Version |
| --- | --- |
| Firefly III | 5.7.6 |
| Firefly III API | 1.5.6 |
| Build | 185, 18-05-2022 06:54:00 CEST |
| DB version | 18 (exp. 18) |
| PHP | `8.0.17` |
| Host | `Linux` |

| System info | Value |
| --- | --- |
| System TZ | America/Toronto |
| Browser TZ | America/Toronto |
| App environment | local |
| App debug mode | true |
| App cache driver | file |
| App logging | debug, stack |
| Display errors | Off |
| Error reporting | ALL errors |
| Interface | apache2handler |
| Default language | en_US |
| Default locale | equal |
| BCscale | 24 |
| DB drivers | mysql, pgsql, sqlite |
| Current driver | mysql |
| Trusted proxies (.env) | `**` |

| User info | Value |
| --- | --- |
| Login provider | eloquent  |
| User guard | web  |
| Headers | N/A, N/A  |
| Stateful domains |  |
| Session start | 2022-05-01 00:00:00 |
| Session end | 2022-05-31 23:59:59 |
| Session first | 2022-04-01 00:00:00 |
| User ID | 1 |
| User language | en_US |
| User locale | en_US |
| Attempt at "en_US.utf8" | 'en_US.utf8' |
| Attempt at "en_US.UTF-8" | 'en_US.UTF-8' |
| User agent | Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.162 Safari/537.36 |```


### Steps to reproduce

_No response_

### Additional info

_No response_

**Comment by JC5** [comment]
I can't tell from your examples, but I'm getting the impression that your original transaction's description has multiple spaces between the words?

**Comment by JC5** [comment]
Ah yes, I can see the spaces in the edit-version of your post.

The search engine will indeed truncate multiple space in a query, and may also trim spaces from the beginning or end of the query.

If you have transactions that have multiple spaces in the description or in an account name you could use "description_contains" or "source_account_contains".

**Reply by sherif-fanous** [reply]
@JC5 Thanks for the replies.

Yes, the CSV file generated from my institution has some 'space' quirks.

Transaction descriptions either have multiple trailing spaces or there are multiple consecutive spaces between the last and 2nd to last words.

So is the recommendation to pre-process the CSV file to cleanse it up before importing?

**Reply by JC5** [reply]
Or use the suggested rule options, that seems easier. :)
